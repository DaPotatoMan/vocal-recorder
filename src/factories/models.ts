import { nanoid } from 'nanoid'
import { type AudioCodec, getBlobCodec } from '../encoder'
import { AudioPeaks, getAudioInfo } from './peaks'

export class Duration extends Number {
  constructor(ms: number) { super(ms) }
  get seconds() { return this.valueOf() / 1000 }
  get minutes() { return this.seconds / 60 }
  get hours() { return this.seconds / 3600 }
}

export class AudioBlob extends Blob {
  /** UID of AudioBlob instance. Automatically set on init */
  id = nanoid()

  constructor(
    public blob: Blob,
    public duration: Duration,
    public peaks: AudioPeaks,
    public codec: AudioCodec
  ) {
    super([blob], { type: blob.type })
    Object.freeze(this)
  }

  toSerialized() {
    return {
      id: this.id,
      blob: this.blob,
      duration: this.duration.valueOf(),
      peaks: Array.from<number>(this.peaks),
      codec: this.codec
    }
  }

  /** Returns filename with proper extension of codec */
  getFilename = (name: string) => `${name}.${this.codec.extension}` as const

  /**
   * @param blob Audio blob
   * @param duration in milliseconds
   */
  static fromRaw(blob: Blob, duration: number, peaksArgs: ConstructorParameters<typeof AudioPeaks>, codec: AudioCodec) {
    return new AudioBlob(blob, new Duration(duration), new AudioPeaks(...peaksArgs), codec)
  }

  /** Create AudioBlob from {@link Blob} */
  static async fromFile(file: Blob) {
    const { peaks, duration } = await getAudioInfo(file)
    const codec = getBlobCodec(file)

    return AudioBlob.fromRaw(file, duration, [peaks], codec as any)
  }
}

export type SerializedAudioBlob = ReturnType<AudioBlob['toSerialized']>
