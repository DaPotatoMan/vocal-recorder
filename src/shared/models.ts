import { AudioPeaks, getAudioInfo } from './peaks'

/** Duration in milliseconds */
export class Duration extends Number {
  get seconds() { return this.valueOf() / 1000 }
  get minutes() { return this.seconds / 60 }
  get hours() { return this.seconds / 3600 }

  toJSON = () => this.valueOf()
}

export class AudioBlob extends Blob {
  constructor(
    public blob: Blob,
    public duration: Duration,
    public peaks: AudioPeaks
  ) {
    super([blob], { type: blob.type })
    Object.freeze(this)
  }

  /** Returns filename with proper extension of codec */
  getFilename = (name: string) => `${name}.mp3` as const

  toJSON() {
    return {
      blob: this.blob,
      duration: this.duration.valueOf(),
      peaks: Array.from<number>(this.peaks)
    }
  }

  static fromJSON(json: AudioBlob.Serialized) {
    return new AudioBlob(json.blob, new Duration(json.duration), new AudioPeaks(json.peaks))
  }

  /** Create AudioBlob from {@link Blob} or {@link ArrayBuffer} */
  static async parse(input: Blob | ArrayBuffer) {
    const { peaks, duration } = await getAudioInfo(input)
    const blob = input instanceof Blob ? input : new Blob([input], { type: 'audio/mpeg' })

    return new AudioBlob(blob, new Duration(duration), new AudioPeaks(peaks))
  }
}

export namespace AudioBlob {
  export type Serialized = ReturnType<AudioBlob['toJSON']>
}
