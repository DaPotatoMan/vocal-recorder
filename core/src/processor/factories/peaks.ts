function expandArray(source: number[], newSize: number): number[] {
  const originalSize = source.length
  const step = (originalSize - 1) / (newSize - 1) // Calculate the step size

  const newArray: number[] = []

  for (let i = 0; i < newSize; i++) {
    const index = i * step
    const lowerIndex = Math.floor(index)
    const upperIndex = Math.ceil(index)

    if (lowerIndex === upperIndex) {
      // The index is on an existing peak, just copy the value
      newArray.push(source[lowerIndex])
    }
    else {
      // Perform linear interpolation between neighboring peaks
      const interpolationFactor = index - lowerIndex
      const lowerValue = source[lowerIndex]
      const upperValue = source[upperIndex]
      const interpolatedValue = lowerValue + (upperValue - lowerValue) * interpolationFactor
      newArray.push(interpolatedValue)
    }
  }

  return newArray
}

function getArrayInRange(source: number[], toMin: number, toMax: number): number[] {
  const min = Math.min(...source)
  const max = Math.max(...source)
  const range = max - min

  // Calculate the normalized value for each element in the array
  return source.map((num) => {
    // Avoid division by zero and handle edge cases
    if (range === 0)
      return toMin + (toMax - toMin) / 2 // Return the midpoint of the custom range

    // normalized value
    return toMin + ((num - min) / range) * (toMax - toMin)
  })
}

function getArrayAverage(source: number[], samples = 64) {
  // Normalize props
  samples = Math.floor(samples)

  const blockSize = Math.floor(source.length / samples)
  const filteredData = []

  let sum = 0

  for (let i = 0; i < samples; i++) {
    const blockStart = blockSize * i // the location of the first sample in the block

    for (let j = 0; j < blockSize; j++)
      sum += Math.abs(source[blockStart + j]) // find the sum of all the samples in the block

    filteredData.push(sum / blockSize) // divide the sum by the block size to get the average
    sum = 0
  }

  // Normalizes the audio data to make a cleaner illustration
  const maxPeak = Math.max(...filteredData)

  // Return raw data if fully silent
  if (maxPeak <= 0.0001)
    return filteredData

  const multiplier = maxPeak ** -1
  return filteredData.map(n => n * multiplier * 100)
}

export class AudioPeaks extends Array {
  constructor(source: number[], public readonly samples = 100, public readonly rangeMin = 0, public readonly rangeMax = 100) {
    source ||= []

    // Expand data if required
    if (samples > source.length)
      source = expandArray(source, samples)

    source = getArrayAverage(source, samples)
    source = getArrayInRange(source, rangeMin, rangeMax)

    super(...source)
  }

  clone(samples = this.samples, rangeMin = this.rangeMin, rangeMax = this.rangeMax) {
    return new AudioPeaks(this, samples, rangeMin, rangeMax)
  }

  getAverage(samples = this.samples) {
    return getArrayAverage(this, samples)
  }

  getToRange(min: number, max: number) {
    return getArrayInRange(this, min, max)
  }
}
