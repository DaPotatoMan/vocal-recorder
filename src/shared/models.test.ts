import { AudioBlob, Duration } from './models'
import { AudioPeaks } from './peaks'

describe('class: AudioBlob', () => {
  const result = new AudioBlob(
    new Blob(),
    new Duration(0),
    new AudioPeaks([])
  )

  const serialized = result.toSerialized()
  serialized.id = 'test-id'

  it('serializes properly', () => {
    expect(serialized).toMatchSnapshot()
  })
})
