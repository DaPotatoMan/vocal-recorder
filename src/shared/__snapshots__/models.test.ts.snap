// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`class: AudioBlob > serializes properly 1`] = `
{
  "blob": Blob {
    Symbol(kHandle): Blob {},
    Symbol(kLength): 0,
    Symbol(kType): "",
  },
  "duration": 0,
  "peaks": [
    NaN,
    NaN,
    NaN,
    NaN,
    NaN,
    NaN,
    NaN,
    NaN,
    NaN,
    NaN,
    NaN,
    NaN,
    NaN,
    NaN,
    NaN,
    NaN,
    NaN,
    NaN,
    NaN,
    NaN,
    NaN,
    NaN,
    NaN,
    NaN,
    NaN,
    NaN,
    NaN,
    NaN,
    NaN,
    NaN,
    NaN,
    NaN,
    NaN,
    NaN,
    NaN,
    NaN,
    NaN,
    NaN,
    NaN,
    NaN,
    NaN,
    NaN,
    NaN,
    NaN,
    NaN,
    NaN,
    NaN,
    NaN,
    NaN,
    NaN,
    NaN,
    NaN,
    NaN,
    NaN,
    NaN,
    NaN,
    NaN,
    NaN,
    NaN,
    NaN,
    NaN,
    NaN,
    NaN,
    NaN,
    NaN,
    NaN,
    NaN,
    NaN,
    NaN,
    NaN,
    NaN,
    NaN,
    NaN,
    NaN,
    NaN,
    NaN,
    NaN,
    NaN,
    NaN,
    NaN,
    NaN,
    NaN,
    NaN,
    NaN,
    NaN,
    NaN,
    NaN,
    NaN,
    NaN,
    NaN,
    NaN,
    NaN,
    NaN,
    NaN,
    NaN,
    NaN,
    NaN,
    NaN,
    NaN,
    NaN,
  ],
}
`;

exports[`class: Duration > serializes properly 1`] = `3600000`;
